File: Partition_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode *lowbegin = NULL, *highbegin = NULL,*low = NULL, *high = NULL, *cur = head;
        
        while(cur)
        {
            if(cur->val < x)
            {
                if (low == NULL)
                    lowbegin = low = new ListNode(cur->val);
                else
                {
                    low->next = new ListNode(cur->val);
                    low = low->next;
                }
            }
            else
            {
                if(high == NULL)
                    highbegin = high = new ListNode(cur->val);
                else
                {
                    high->next = new ListNode(cur->val);
                    high = high->next;
                }
            }
            cur = cur->next;
        }
        if(low == NULL || high == NULL)
            return head;
        else
        {
            low->next = highbegin;
            return lowbegin;
        }
    }
};

File: leetcode/twopointer/Partition List.go
package twopointer

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func partition(head *ListNode, x int) *ListNode {
	d1, d2 := &ListNode{0, nil}, &ListNode{0, nil}
	c1, c2 := d1, d2
	c := head
	for; c != nil;{
		if c.Val < x{
			c1.Next = c
			c1 = c1.Next
		}else{
			c2.Next = c
			c2 = c2.Next
		}
		c = c.Next
	}
	c1.Next = d2.Next
	c2.Next = nil
	return d1.Next
}